content/posts/cicd_ha.md

```diff
@@ -0,0 +1,132 @@
+
+### Resilient vs Highly Available
+
+Once said that, one of the most important features about a *Continuous Delivery* system (note that I am talking in general, not only software) to be highly available, is **resilience**. Understood as the capacity of our system components to recover on time, producing a complete stable, durable and efficient execution that doesn't impact the overall service (e.g. my deployment or release frequency and quality of final artifacts are the same, with or without short downtimes. Just because my platform and systems are *resilient* to failures).
+
+Let's back then to the question:
+
+> ***What platform configuration do I need for a highly available service?***
+
+To answer this question, it's very important to think about **resilience first**, because having a resilient platform and architecture is usually enough to deliver a highly available service, being able to recover on time to not impact the final service. But in some scenarios where a couple of minutes of service blockage means loss of millions of dollars (like Amazon retail previous example), resilient solutions are not enough. In this case then you probably need to work on very restrictive HA replication infrastructures. The cost of replication complexity need to be worth it.
+
+## So CI/CD HA is just Resilience
+
+Coming back to main topic... **YES**, it is about resilient solutions. HA in CI/CD is not about looking for an *active-active* load balancing platform that is able to orchestrate CI/CD pipelines. It is about assuring enough *availability* of my pipelines executions to deliver my software on time and covering all expectations (deployment frequency, failure recovery, lead time, quality standards, etc.).
+
+What does resilient architecture mean in CI/CD pipeline orchestration? I think that following solution architecture concepts are a must in order to offer a resilient *highly available* solution:
+
+* Orchestration layer does not have the same needs that execution layer. Pipeline orchestrators and pipeline executors have different scalability  requirements, and because of that **decoupling execution from orchestration is a must**.
+* **Pipeline execution downtime recovery needs to be fast** and from the same execution point. Executors infrastructure can fail, even if they are replicated, so it is more important to recover from the last successful stage or status than moving the execution to another live agent executor that could restart the execution from the beginning
+* **Decouple also the orchestration layer**. Smaller and more masters (orchestrators) are going to be more scalable and resilient to any overall failure. If one master fails it is not impacting any other orchestrator and its pipelines. So platform availability improves with scalability.
+* Rely on a **native self-healing, resilient and scalable infrastructure**. Event thought that replication is not needed, we need to recover fast from any hardware or service downtime.
```

Comment:
"Event thought that replication is not needed"
to
"Even though replication is not needed"